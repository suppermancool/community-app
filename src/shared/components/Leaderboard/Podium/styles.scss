@import '~styles/mixins';

.Podium {
  @include roboto-regular;

  line-height: 1.42;
  margin: 20px -15px;
  text-align: center;

  &::before,
  &::after {
    content: " ";
    display: table;
  }

  &::after {
    clear: both;
  }
}

.Podium .podium-column {
  &.full {
    width: 100%;
  }
  width: 33.33333%;
  float: left;
  padding: 0 15px;
}
