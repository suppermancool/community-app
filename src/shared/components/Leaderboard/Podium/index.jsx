/**
 * Leaderboard Podium
 *
 * Description:
 *   Component that display the top spots on a leaderboard
 *   It is designed to work with the PodiumSpot component. The Podium component
 *   simply created a layout to display PodiumSpots.
 *
 * Usage:
 *   <Podium competitors={competitors} />
 *
 * Props:
 *   - competitors (required): Array of Top Coder user object, with
 *     the following properties:
 *      - rank: Number, required. User current ranking in the leaderboard
 *      - photourl: String. URL for the user's profile picture
 *        This will default to the default user avatar is undefined
 *      - user.handle: String, required. User handle
 *      - challenge.count: Number, required. The number of challenge the user won
 *      - project_result.final_score: Number, required. The user's current score
 */

import React from 'react';
import PT from 'prop-types';
import LoadingIndicator from 'components/LoadingIndicator';
import PodiumSpot from '../PodiumSpot';

import './styles.scss';

export default function Podium(props) {
  const {
    competitors,
  } = props;

  if (competitors.length > 1) {
    const comp0 = competitors[0];
    const comp1 = competitors[1];
    competitors[0] = comp1;
    competitors[1] = comp0;
  }

  const renderPodium = (comps) => {
    if (comps.length === 0) {
      return (
        <LoadingIndicator />
      );
    }

    const renderItems = items => (
      items.map(competitor => (
        <div styleName={`podium-column ${(items.length === 1) ? 'full' : ''}`}>
          <PodiumSpot competitor={competitor} />
        </div>
      ))
    );

    return (
      <div>
        {renderItems(comps)}
      </div>
    );
  };

  return (
    <div styleName="Podium">
      {renderPodium(competitors)}
    </div>
  );
}

const CompetitorShape = PT.shape({
  rank: PT.number.isRequired,
  'challenge_stats.photo_url': PT.string,
  'challenge_stats.winner_handle': PT.string.isRequired,
  'challenge_stats.count': PT.number.isRequired,
  points: PT.number.isRequired,
});

Podium.propTypes = {
  competitors: PT.arrayOf(CompetitorShape).isRequired,
};
