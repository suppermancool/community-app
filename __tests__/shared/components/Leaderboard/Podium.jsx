import React from 'react';
import Renderer from 'react-test-renderer/shallow';
import Podium from 'components/Leaderboard/Podium';

const mockUserObj = {
  rank: 1,
  photourl: 'some.fake.string',
  handle: 'overachieving.mofo',
  challengecount: 99999,
  points: 9999999999,
};

const mockLeaderboardData = [mockUserObj, mockUserObj, mockUserObj, mockUserObj];

test('Matches shallow shapshot', () => {
  const renderer = new Renderer();
  renderer.render((
    <Podium competitors={mockLeaderboardData} />
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();

  renderer.render((
    <Podium competitors={[]} />
  ));
  expect(renderer.getRenderOutput()).toMatchSnapshot();
});
