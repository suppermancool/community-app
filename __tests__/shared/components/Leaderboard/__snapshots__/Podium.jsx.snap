// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches shallow shapshot 1`] = `
<div
  className="src-shared-components-Leaderboard-Podium-___styles__Podium___2rb4T"
>
  <div>
    <div
      className="src-shared-components-Leaderboard-Podium-___styles__podium-column___rPHl-"
    >
      <PodiumSpot
        competitor={
          Object {
            "challengecount": 99999,
            "handle": "overachieving.mofo",
            "photourl": "some.fake.string",
            "points": 9999999999,
            "rank": 1,
          }
        }
      />
    </div>
    <div
      className="src-shared-components-Leaderboard-Podium-___styles__podium-column___rPHl-"
    >
      <PodiumSpot
        competitor={
          Object {
            "challengecount": 99999,
            "handle": "overachieving.mofo",
            "photourl": "some.fake.string",
            "points": 9999999999,
            "rank": 1,
          }
        }
      />
    </div>
    <div
      className="src-shared-components-Leaderboard-Podium-___styles__podium-column___rPHl-"
    >
      <PodiumSpot
        competitor={
          Object {
            "challengecount": 99999,
            "handle": "overachieving.mofo",
            "photourl": "some.fake.string",
            "points": 9999999999,
            "rank": 1,
          }
        }
      />
    </div>
  </div>
</div>
`;

exports[`Matches shallow shapshot 2`] = `
<div
  className="src-shared-components-Leaderboard-Podium-___styles__Podium___2rb4T"
>
  <ThemedLoadingIndicator
    composeAdhocTheme="deeply"
    composeContextTheme="softly"
    mapThemrProps={[Function]}
    themePriority="adhoc-context-default"
  />
</div>
`;
